File: DataStructures/sorts/go/quick.go
package main

func swap(list *[]int, i, j int) {
	temp := (*list)[i]
	(*list)[i] = (*list)[j]
	(*list)[i] = temp
}

func partition(list *[]int, start, end int) int {
	left := start - 1
	pivot := (*list)[end]

	for i := start; i < end; i++ {
		if (*list)[i] < pivot {
			left++
			swap(list, i, left)
		}
	}
	swap(list, left+1, end)
	return left + 1
}

func quickSort(list *[]int, start, end int) {
	if start < end {
		mid := partition(list, start, end)
		quickSort(list, start, mid-1)
		quickSort(list, mid+1, end)
	}
}

File: DataStructures/sorts/js/quick.js
function swap(list, i, j) {
    let temp = list[i]
    list[i] = list[j]
    list[j] = temp
}

function partition(list, start, end) {
    const pivot = list[end];
    let leftPointer = start - 1;
    for (let i = start; i < end; ++i) {
        if (list[i] < pivot) {
            leftPointer++;
            swap(list, leftPointer, i)
        }
    }
    swap(list, leftPointer + 1, end);
    return leftPointer + 1;
}

function quickSort(list, start = 0, end = list.length - 1) {
    if (start < end) {
        const pivot = partition(list, start, end)
        quickSort(list, start, pivot - 1)
        quickSort(list, pivot + 1, end)
    }
    return list
}

